include/options.hxx

```diff
@@ -158,13 +161,76 @@ public:
   /// @param[in] sectionName   Name of the section, including path from the root
   Options(Options *parent_instance, std::string full_name)
       : parent_instance(parent_instance), full_name(std::move(full_name)){};
-  
+
+  /// Copy constructor
+  Options(const Options& other);
+
   /// Get a reference to the only root instance
   static Options &root();
   
   /// Free all memory
   static void cleanup();
 
+  /// The type used to store values
+  using ValueType =
+      bout::utils::variant<bool, int, BoutReal, std::string, Field2D, Field3D,
+                           Array<BoutReal>, Matrix<BoutReal>, Tensor<BoutReal>>;
+
+  /// The type used to store attributes
+  /// Extends the variant class so that cast operator can be implemented
+  /// and assignment operator overloaded
+  ///
+  /// Note: Due to default initialisation rules, if an attribute
+  /// is used without being set, it will be false, 0, 0.0 and
+  /// throw std::bad_cast if cast to std::string
+  /// 
+  class AttributeType : public bout::utils::variant<bool, int, BoutReal, std::string> {
+  public:
+    using Base = bout::utils::variant<bool, int, BoutReal, std::string>;
+
+    /// Constructor
+    AttributeType() {}
+    /// Copy constructor
+    AttributeType(const AttributeType& other) : Base(other) {}
+    /// Move constructor
+    AttributeType(AttributeType&& other) : Base(other) {}
```

Comment:
I think this should this be `Base(std::move(other))` as `other` becomes an l-value in the initialiser list